pkg/sysconfwatch/sysconfwatch.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright © 2022 Microshift Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sysconfwatch
+
+import (
+	"context"
+	"math"
+	"os"
+	"time"
+
+	"github.com/openshift/microshift/pkg/config"
+	"github.com/openshift/microshift/pkg/util"
+	"golang.org/x/sys/unix"
+	"k8s.io/klog/v2"
+)
+
+const sysConfigCheckInterval = time.Second * 5
+
+type SysConfWatchController struct {
+	NodeIP  string
+	timerFd int
+}
+
+func NewSysConfWatchController(cfg *config.MicroshiftConfig) *SysConfWatchController {
+	// Create a realtime clock timer with asynchronous read support
+	fd, err := unix.TimerfdCreate(unix.CLOCK_REALTIME, unix.TFD_CLOEXEC|unix.TFD_NONBLOCK)
+	if err != nil {
+		klog.Fatalf("failed to create a realtime clock timer %v", err)
+	}
+
+	// Set the time interval into distant future
+	var ptime = &unix.ItimerSpec{
+		Interval: unix.Timespec{Sec: math.MaxInt64, Nsec: 0},
+		Value:    unix.Timespec{Sec: 0, Nsec: 0},
+	}
+	// Start the timer with cancelation if the clock is reset elsewhere
+	err = unix.TimerfdSettime(fd, unix.TFD_TIMER_ABSTIME|unix.TFD_TIMER_CANCEL_ON_SET, ptime, nil)
+	if err != nil {
+		klog.Fatalf("failed to start a realtime clock timer %v", err)
+	}
+
+	return &SysConfWatchController{
+		NodeIP:  cfg.NodeIP,
+		timerFd: fd,
+	}
+}
+
+func (s *SysConfWatchController) Name() string { return "sysconfwatch-controller" }
+func (s *SysConfWatchController) Dependencies() []string {
+	return []string{}
+}
+
+func (c *SysConfWatchController) Run(ctx context.Context, ready chan<- struct{}, stopped chan<- struct{}) error {
+	defer close(stopped)
+	ticker := time.NewTicker(sysConfigCheckInterval)
+	defer ticker.Stop()
+
+	klog.Infof("starting sysconfwatch-controller with IP address %q", c.NodeIP)
+
+	var buf []byte = make([]byte, 8)
+	for {
+		select {
+		case <-ticker.C:
+			// Check the IP change
+			currentIP, _ := util.GetHostIP()
+			if c.NodeIP != currentIP {
+				klog.Warningf("IP address has changed from %q to %q, restarting MicroShift", c.NodeIP, currentIP)
+				os.Exit(0)
+				return nil
+			}
+
+			// Check the clock change
+			// Initiate an asynchronous read operation on the timer object
+			// When the clock is reset, the read operation returns with the ECANCELED error code
+			_, err := unix.Read(c.timerFd, buf)
+			if err == unix.ECANCELED {
+				klog.Warningf("realtime clock change detected, restarting MicroShift")
```

Comment:
I believe we should validate the clock change to see how much it was.

For example, an ntp clock sync would trigger this, but the change would be in the range of <1s, which may not trigger a restart of MicroShift. (Otherwise we get a restart every time ntp sync happens).

I suggest we:
1)  store the clock time at start, 
2) If the event happens: calculate the diff, if <"1-5s"  we log, and ignore, also store the new time for next check
3) If the diff from step 2 is >1-5s then we log, and trigger the restart.

Does it sound reasonable @ggiguash ?

Reply:
Yes, we need to avoid that situation when NTP fires. The only thing to check is whether OCP modules may "break" on smaller 1-5s time update.

Reply:
@mangelajo, I looked into reliable calculation of the time difference between two samples. While in principle your suggestion works, it does not seem to be fail-proof. Because we are comparing 2 time samples, we need to guarantee that the clock update happened after the first sample was taken. I do not think there is a way to achieve this reliably. 

Working sequence
1. Take time sample 1
2. Sleep 5 secs
<Time is adjusted>
3. Time change event received: Take time sample 2
4. Check the time diff and decide if to exit
5. Next iteration

Problematic sequence is when the time adjustment occurs on step 5 for example. Then, the sample 1 contains the adjusted time, event is received, but sample 2 will give a valid range of time difference.

Reply:
@mangelajo, I uploaded an implementation of time drift calculation, taking into account changes between system and monotonic clock offsets. The current allowed time drift is set to +/- 10 seconds.
If you approve of this code, resolve the conversation and I will update the docs.